client/with_dry_runs.go

```diff
@@ -0,0 +1,236 @@
+package client
+
+import (
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/mysteriumnetwork/payments/bindings"
+	"github.com/pkg/errors"
+)
+
+// WithDryRuns forces a dry run before running a write transaction on blockchain.
+// Ethereum client will perform a dry run on a transaction with no gas limit set.
+// This component will perform a dry run if and only if the gas limit is set to a non zero value.
+// In this way, the dry run is always performed before sending the transaction to the network.
+// For convenience, this component proxies read only calls to the underlying blockchain.
+type WithDryRuns struct {
+	bc blockchain
+}
+
+// NewWithDryRuns creates a new instance of client with dry runs.
+func NewWithDryRuns(bc blockchain) *WithDryRuns {
+	return &WithDryRuns{
+		bc: bc,
+	}
+}
+
+type gasLimitProvider interface {
+	GetGasLimit() uint64
+}
+
+func (cwdr *WithDryRuns) dryRun(glp gasLimitProvider, binding string, from, to common.Address, method string, params ...interface{}) (uint64, error) {
```

Comment:
Maybe add comment that dryRun params should be passed in strict order as defined in smart contracts?